Firmware/MotorControl/pwm_input.cpp

```diff
@@ -0,0 +1,423 @@
+        } break;
+        case PWM_TYPE_DUTY_CYCLE: {
+            if (high_time > period)
+                return false;
+            ratio = (float)high_time / (float)period;
+        } break;
+        default: {
+            return true;
+        } break;
+    }
+
+    if (board_config.pwm_mappings[gpio_index].enable_deadband) {
+        ratio = add_deadband(ratio);
+    }
+
+    set_channel_output(endpoint, gpio_index, ratio);
+
+    return true;
+}
+
+// In this function we need to calculate the duty cycle of the incoming PWM signal so that it can be translated into a control input
+
+// Three types of pulse-width modulation (PWM) are possible:
+// 1. The pulse center may be fixed in the center of the time window and both edges of the pulse moved to compress or expand the width.
+// 2. The lead edge can be held at the lead edge of the window and the tail edge modulated.
+// 3. The tail edge can be fixed and the lead edge modulated.
+
+// Is there a good way to check the period of all 3 types?  If not, type 2 should be the most common in this situation.
+// Type 2 is also the RC PWM signal type, so this implementation will work for both.
+
+// How can we verify that a pulse is a valid signal? RC PWM has this built-in, where a pulse needs to be within a certain length, and the period should be around
+// 20ms, but duty-cycle PWM has no such restrictions.  Right after boot, the old code waits for several HtoL transitions before it considers the input valid, but
+// shouldn't this be done every time a signal goes inactive?
+// We could compare the period of each pulse to the previous one, and require one or more consecutive matches for the pulse to be considered valid.  This would
+// work for RC PWM too, and would replace the previous validation method.
+
+// dutyCycle = highTime / period
+
+// Calculating the period with either of these two formulas...
+// period = highTime + lowTime
+// period = HtoL_timestamp - last_HtoL_timestamp
+// ...depends on knowing the order of the high/low pulses in the cycle.  If we assume it's type two, it will be high then low.
+// As long as the duty cycle changes relatively slowly, if it's type 1 or 3 it shouldn't be too inaccurate.
+
+// We also need to handle the case where the PWM signal has no duty cycle, and is simply switching between max and min at random times.
+// We need to put a minimum period requirement on the signal, because there is no other way to differentiate between, say, switching from high to low
+// every 3 seconds, and a 50% PWM signal with a period of 6 seconds.
+
+// Timer CCxOF bit can be checked to see if two captures occurred, the second overwriting the first
+// "In order to handle the overcapture, it is recommended to read the data before the
+//  overcapture flag. This is to avoid missing an overcapture which could happen after reading
+//  the flag and before reading the data."
+// Maybe we can use this to set the output values to min or max depending on current pin state
```

Comment:
The `CCxOF` bit would only help detecting a very short pulse, but not no pulse at all. Some form of timeout monitoring is required for that.

Reply:
That's what I was using it for.  Timeout monitoring was being done in pwm_no_pulse_check_cb().